### PR TITLE
Multipart mime 3487 v9

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3487

Describe changes:
- convert HTTP to use new rust mime parser
- convert SMTP to use new rust mime parser

Follows #8299 with needed rebase

Still to do : review !
Should I list my questions ?

suricata-verify-pr: 1035

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | build_asan | | |
 

Pipeline 11302

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW1_files_sha256.
 | field | baseline | test | % |
|--|--|--|--| 
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 120905 | 0.02 | 

Pipeline 11305

**Comment by inashivb** [comment]
This is super awesome. Thank you for this! :) ❤️

File: src/app-layer-smtp.c [hunk 1]
```diff
@@ -1263,8 +1151,9 @@ static int SMTPPreProcessCommands(
     DEBUG_VALIDATE_BUG_ON((state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE) == 0);
 
     /* fall back to strict line parsing for mime header parsing */
+    // TODOrust4 use consumed API
```
Comment (catenacyber) [inline src/app-layer-smtp.c:1154]: Should this be part of the PR ?
This being the conversion of SMTP to use the AppLayerResult::Incomplete

File: src/app-layer-smtp.c [hunk 2]
```diff
@@ -734,39 +567,37 @@ static int SMTPProcessCommandBDAT(
...
-    if (msg->anomaly_flags & ANOM_MALFORMED_MSG) {
+    if (events & MIME_ANOM_MALFORMED_MSG) {
         SMTPSetEvent(state, SMTP_DECODER_EVENT_MIME_MALFORMED_MSG);
     }
-    if (msg->anomaly_flags & ANOM_LONG_BOUNDARY) {
+    if (events & MIME_ANOM_LONG_BOUNDARY) {
         SMTPSetEvent(state, SMTP_DECODER_EVENT_MIME_BOUNDARY_TOO_LONG);
     }
-    if (msg->anomaly_flags & ANOM_LONG_FILENAME) {
+    if (events & MIME_ANOM_LONG_FILENAME) {
```
Comment (catenacyber) [inline src/app-layer-smtp.c:600]: Do we still need these ?

File: rust/src/mime/smtp.rs [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+    if let Some(x) = input.iter().position(|&x| {
+        x == b' ' || x == b'"' || x == b'\'' || x == b'<' || x == b'>' || x == b']' || x == b'\t'
+    }) {
+        return &input[..x];
+    }
+    return input;
+}
+
+fn mime_smtp_extract_urls(urls: &mut Vec<Vec<u8>>, input_start: &[u8]) {
+    //TODOrust6 use mpm
```
Comment (catenacyber) [inline rust/src/mime/smtp.rs:259]: thoughts about this optimization ? (using a mpm)

File: src/util-decode-mime.c [hunk 4]
```diff
@@ -1,3580 +0,0 @@
...
-
-/**
- * \brief Pushes a node onto a stack and returns the new node.
- *
- * \param stack The top of the stack
- *
- * \return pointer to a new node, otherwise NULL if it fails
- */
-static MimeDecStackNode * PushStack(MimeDecStack *stack)
-{
```
Comment (catenacyber) [inline src/util-decode-mime.c:463]: What feature do we need for mime inside mime ?

File: src/util-decode-mime.c [hunk 5]
```diff
@@ -1,3580 +0,0 @@
...
-
-    if (state->body_end) {
-        SCLogDebug("Body end (len=%u)", len);
-    }
-
-    return MIME_DEC_OK;
-}
-
-/* Test simple case of line counts */
-static int MimeDecParseLineTest01(void)
```
Comment (catenacyber) [inline src/util-decode-mime.c:2690]: Should I translate these tests to rust ?

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/8384

File: src/util-decode-mime.c (hunk 4 above)
Comment (victorjulien) [reply src/util-decode-mime.c:463]: I don't know, same as with C?

File: rust/src/mime/smtp.rs (hunk 3 above)
Comment (victorjulien) [reply rust/src/mime/smtp.rs:259]: seems nice but can be done later?

File: src/app-layer-smtp.c (hunk 2 above)
Comment (victorjulien) [reply src/app-layer-smtp.c:600]: why / why not?

File: src/app-layer-smtp.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-smtp.c:1154]: I don't understand the question.

Comment (catenacyber) [reply src/app-layer-smtp.c:1154]: Should the conversion of SMTP to use `AppLayerResult::Incomplete` be part of this PR that translates mime into rust ?

File: src/util-decode-mime.c (hunk 4 above)
Comment (catenacyber) [reply src/util-decode-mime.c:463]: As far as I get the C version, there seems to be complex stuff that are done and then unused...

File: src/app-layer-smtp.c (hunk 2 above)
Comment (catenacyber) [reply src/app-layer-smtp.c:600]: This zen answer tells me that you do not need these if you do not know if you need them ;-p

Comment (victorjulien) [reply src/app-layer-smtp.c:600]: Confusing. If we need them depends on where they are used, I think.

File: src/app-layer-smtp.c (hunk 1 above)
Comment (catenacyber) [reply src/app-layer-smtp.c:1154]: I think not, but I will check... Is this supposed to be a recent change ?

Comment (catenacyber) [reply src/app-layer-smtp.c:1154]: `SMTPPreProcessCommands` does not use `AppLayerResult::Incomplete` but `SMTPGetLine` does

So, it is only in the SMTP headers parsing, not in the DATA mode...

File: src/app-layer-smtp.c (hunk 2 above)
Comment (catenacyber) [reply src/app-layer-smtp.c:600]: They are used to set events : anomaly and rules with `app-layer-event`

Putting them

**Pull request closed without merging**